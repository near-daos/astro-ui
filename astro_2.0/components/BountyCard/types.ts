import { Token } from 'types/token';
import { BountyStatus } from 'types/bounties';

export enum CardType {
  Bounty = 'Bounty',
  Claim = 'Claim',
}
export interface BountyCardContent {
  id: string;
  daoId: string;
  amount: string;
  type: CardType;
  description: string;
  externalUrl?: string;
  token: Token;
  status: BountyStatus;
  bountyBond: string;
  slots?: number;
  slotsTotal?: number;
  claimedByCurrentUser: boolean;
  accountId?: string;
  timeToComplete: string;
  forgivenessPeriod: string;
  relatedProposalHash?: string;
}
