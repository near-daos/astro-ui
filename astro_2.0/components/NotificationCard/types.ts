import { NotificationType, NotificationStatus } from 'types/notification';

export interface NotificationCardContent {
  id: string;
  type: NotificationType;
  status: NotificationStatus;
  text: string;
  time: string;
  flagCover?: string;
  logo?: string;
  url?: string;
}
