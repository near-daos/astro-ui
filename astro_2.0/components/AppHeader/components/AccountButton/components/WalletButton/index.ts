export * from 'astro_2.0/components/AppHeader/components/AccountButton/components/WalletButton/WalletButton';
