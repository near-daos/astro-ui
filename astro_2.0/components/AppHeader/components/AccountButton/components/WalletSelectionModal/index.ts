export * from 'astro_2.0/components/AppHeader/components/AccountButton/components/WalletSelectionModal/WalletSelectionModal';
