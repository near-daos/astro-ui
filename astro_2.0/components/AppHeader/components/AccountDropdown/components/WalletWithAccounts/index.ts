export * from 'astro_2.0/components/AppHeader/components/AccountDropdown/components/WalletWithAccounts/WalletWithAccounts';
