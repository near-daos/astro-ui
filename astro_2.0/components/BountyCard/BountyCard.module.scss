@import 'styles/mixins';
@import 'styles/typography';

.root {
  max-width: 462px;
  position: relative;
  width: 100%;
}

.cover {
  @extend %title3;
  background-image: linear-gradient(#201f1f, transparent);
  background-size: cover;
  height: 100%;
  inset: 0;
  width: 50px;
  z-index: -1;
}

.bountyGrid {
  background-color: var(--color-white);
  border-radius: 8px;
  box-shadow: 0 8.89583px 30px var(--color-black-opacity-14);
  display: grid;
  grid-row-gap: 22px;
  grid-template-areas:
    'proposalType daysLeft daysLeft'
    'status status .'
    'description description description'
    'content content content'
    'actions actions actions';
  grid-template-columns: 1fr 1fr 1fr;
  padding: 22px 24px 0;
  position: relative;

  &:hover {
    box-shadow: 0 16px 54px var(--color-black-opacity-25);
  }
}

.completeDate {
  @extend %title4;
  color: var(--color-brand-purple);
  grid-area: daysLeft;
  justify-self: end;
}

.statusAvailable {
  color: var(--color-primary-50);
}

.statusExpired {
  color: var(--color-brand-coral-red);
}

.statusInProgress {
  color: var(--color-success-60);
}

.pendingApproval {
  color: var(--color-blue-50);
}

.proposalType {
  grid-area: proposalType;
}

.status {
  grid-area: status;
}

.countdown {
  grid-area: countdown;
}

.description {
  grid-area: description;
}

.content {
  display: flex;
  grid-area: content;
  justify-content: space-between;
}

.loaderWrapper {
  align-items: flex-end;
  display: flex;
  padding: 8px;
}

.footerLabel {
  @extend %caption1;
  color: var(--color-neutral-60);
}

.proposalLink {
  @extend %body2;

  align-items: center;
  color: var(--color-brand-purple);
  display: flex;
  margin-top: 8px;

  .icon {
    height: 16px;
    margin-right: 8px;
    width: 16px;
  }
}

.slotsTotal {
  @extend %subtitle5;

  color: var(--color-neutral-60);
}
