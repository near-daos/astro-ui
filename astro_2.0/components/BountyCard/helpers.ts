import { format, formatDistance, isAfter } from 'date-fns';

import {
  BountyCardContent,
  CardType,
} from 'astro_2.0/components/BountyCard/types';
import { Tokens } from 'astro_2.0/features/CustomTokens/CustomTokensContext';
import { DAO } from 'types/dao';
import { Bounty, BountyStatus, ClaimedBy } from 'types/bounties';

export const toMillis = (timePeriod: string): number =>
  Math.round(Number(timePeriod) / 1000000);

const calculateBountyStatus = (claim: ClaimedBy): BountyStatus => {
  return isAfter(Date.now(), toMillis(claim.endTime))
    ? BountyStatus.Expired
    : BountyStatus.InProgress;
};

export const getDistanceFromNow = (timePeriod: string): string => {
  return formatDistance(new Date(toMillis(timePeriod)), 0, {
    addSuffix: false,
  });
};

const canClaim = (bounty: Bounty, currentUser: string) => {
  const claimedByMe = bounty.claimedBy
    .map(({ accountId: claimedAccount }) => claimedAccount)
    .includes(currentUser);

  return bounty.slots !== 0 && !claimedByMe;
};

export const formatDate = (date: string): string => {
  return format(toMillis(date), 'd LLLL yyyy');
};

export const mapBountyToCardContent = (
  dao: DAO,
  bounty: Bounty,
  tokens: Tokens,
  currentUser: string,
  filterStatus: BountyStatus
): BountyCardContent[] => {
  const tokenData =
    bounty.tokenId === '' ? tokens.NEAR : tokens[bounty.tokenId];

  const commonProps = {
    id: bounty.id,
    daoId: dao.id,
    amount: bounty.amount,
    description: bounty.description,
    externalUrl: bounty.externalUrl,
    forgivenessPeriod: bounty.forgivenessPeriod,
    token: tokenData,
    bountyBond: dao.policy.bountyBond,
  };

  const result: BountyCardContent[] = [];

  const includeBountyStatuses = [BountyStatus.Available];
  const includeClaimsStatuses = [BountyStatus.InProgress, BountyStatus.Expired];

  if (
    (!filterStatus || includeBountyStatuses.includes(filterStatus)) &&
    canClaim(bounty, currentUser)
  ) {
    result.push({
      ...commonProps,
      status: BountyStatus.Available,
      type: CardType.Bounty,
      timeToComplete: getDistanceFromNow(bounty.deadlineThreshold),
      slots: bounty.slots,
      slotsTotal: bounty.slotsTotal,
      claimedByCurrentUser: false,
    });
  }

  if (!filterStatus || includeClaimsStatuses.includes(filterStatus)) {
    result.push(
      ...bounty.claimedBy.map(claim => {
        return {
          ...commonProps,
          type: CardType.Claim,
          status: calculateBountyStatus(claim),
          claimedBy: claim.accountId,
          claimedByCurrentUser: claim.accountId === currentUser,
          timeToComplete: formatDate(claim.endTime),
        };
      })
    );
  }

  return result;
};

export const showActionBar = (
  bountyCardContent: BountyCardContent,
  account: string | undefined
): boolean => {
  return !!account;
};
