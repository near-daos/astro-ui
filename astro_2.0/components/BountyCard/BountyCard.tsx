import React, { useCallback } from 'react';
import cn from 'classnames';
import Link from 'next/link';
import { FormProvider, useForm } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers/yup';
import * as yup from 'yup';
import { useRouter } from 'next/router';
import { useTranslation } from 'next-i18next';

import { SINGLE_PROPOSAL_PAGE_URL } from 'constants/routing';

import { Proposal } from 'types/proposal';
import { BountyStatus } from 'types/bounties';
import { BountyCardContent } from 'astro_2.0/components/BountyCard/types';
import { DAOFormValues } from 'astro_2.0/features/CreateDao/components/types';
import { DAO } from 'types/dao';

import { Icon } from 'components/Icon';
import { TokenWidget } from 'astro_2.0/components/TokenWidget';
import { InfoBlockWidget } from 'astro_2.0/components/InfoBlockWidget';
import { LoadingIndicator } from 'astro_2.0/components/LoadingIndicator';
import { ProposalDescription } from 'astro_2.0/components/ProposalDescription';
import { CardFooter } from 'astro_2.0/components/BountyCard/components/CardFooter';
import { BountyActionsBar } from 'astro_2.0/components/BountyCard/components/BountyActionsBar';
import { Input } from 'components/inputs/Input';
import { InputWrapper } from 'astro_2.0/features/CreateProposal/components/InputWrapper';

import { gasValidation } from 'astro_2.0/features/CreateProposal/helpers';
import { NOTIFICATION_TYPES, showNotification } from 'features/notifications';
import { SputnikNearService } from 'services/sputnik';

import {
  DEFAULT_PROPOSAL_GAS,
  DEFAULT_VOTE_GAS,
} from 'services/sputnik/constants';

import styles from './BountyCard.module.scss';

const schema = yup.object().shape({
  gas: gasValidation,
});

export interface BountyCardProps {
  bountyId: string;
  deadlineThreshold: string;
  content: BountyCardContent;
  showActionBar: boolean;
  canClaim: boolean;
  dao: DAO;
  relatedProposal?: Proposal;
  completeHandler: () => void;
}

export const BountyCard: React.FC<BountyCardProps> = ({
  bountyId,
  deadlineThreshold,
  content,
  showActionBar,
  canClaim,
  dao,
  completeHandler,
  relatedProposal,
}) => {
  const {
    status,
    description,
    timeToComplete,
    forgivenessPeriod,
    externalUrl,
    token,
    amount,
    bountyBond,
    type,
  } = content;
  const router = useRouter();
  const { t } = useTranslation();
  const daoId = router.query.dao as string;
  const methods = useForm<DAOFormValues>({
    mode: 'all',
    reValidateMode: 'onChange',
    defaultValues: {
      gas: DEFAULT_VOTE_GAS,
    },
    resolver: yupResolver(schema),
  });

  const onSuccessHandler = useCallback(async () => {
    await router.replace(router.asPath);
    showNotification({
      type: NOTIFICATION_TYPES.INFO,
      lifetime: 20000,
      description: t('bountiesPage.successClaimBountyNotification'),
    });
  }, [t, router]);

  const handleClaim = useCallback(
    async data => {
      await SputnikNearService.claimBounty(daoId, {
        bountyId: Number(bountyId),
        deadline: deadlineThreshold,
        bountyBond: dao.policy.bountyBond,
        gas: data.gas,
      });

      onSuccessHandler();
    },
    [
      bountyId,
      dao.policy.bountyBond,
      daoId,
      deadlineThreshold,
      onSuccessHandler,
    ]
  );

  const handleUnclaim = useCallback(
    async data => {
      await SputnikNearService.unclaimBounty(daoId, bountyId, data.gas);
      onSuccessHandler();
    },
    [bountyId, daoId, onSuccessHandler]
  );

  const currentGasValue = methods.watch('gas').toString();

  function getInputWidth() {
    if (currentGasValue?.length > 6 && currentGasValue?.length <= 10) {
      return `${currentGasValue?.length}ch`;
    }

    if (currentGasValue?.length > 10) {
      return '10ch';
    }

    return '6ch';
  }

  function renderActionBar() {
    if (showActionBar) {
      if (status === BountyStatus.PendingApproval && relatedProposal) {
        const { id } = relatedProposal;

        return (
          <CardFooter>
            <div>
              <div className={styles.footerLabel}>Completion Proposal</div>
              <div>
                <Link
                  href={{
                    pathname: SINGLE_PROPOSAL_PAGE_URL,
                    query: {
                      dao: daoId,
                      proposal: id,
                    },
                  }}
                >
                  <a className={styles.proposalLink}>
                    <Icon name="buttonExternal" className={styles.icon} />
                    Link to Proposal
                  </a>
                </Link>
              </div>
            </div>
          </CardFooter>
        );
      }

      return (
        <BountyActionsBar
          canClaim={canClaim}
          bountyBond={bountyBond}
          forgivenessPeriod={forgivenessPeriod}
          bountyStatus={status}
          unclaimHandler={handleUnclaim}
          completeHandler={completeHandler}
        />
      );
    }

    return null;
  }

  return (
    <FormProvider {...methods}>
      <form
        onSubmit={methods.handleSubmit(handleClaim)}
        className={styles.root}
      >
        <div className={styles.bountyGrid}>
          <InfoBlockWidget
            label="Type"
            value={type}
            valueFontSize="L"
            className={styles.proposalType}
          />
          <div className={styles.completeDate}>
            {status === BountyStatus.Available
              ? `Complete in ${timeToComplete}`
              : `Due on ${timeToComplete}`}
          </div>
          <InfoBlockWidget
            label="Status"
            value={
              <div
                className={cn({
                  [styles.statusAvailable]: status === BountyStatus.Available,
                  [styles.statusInProgress]: status === BountyStatus.InProgress,
                  [styles.statusExpired]: status === BountyStatus.Expired,
                  [styles.pendingApproval]:
                    status === BountyStatus.PendingApproval,
                })}
              >
                {status}
              </div>
            }
            valueFontSize="L"
            className={styles.status}
          />
          <ProposalDescription
            description={description}
            link={externalUrl}
            className={styles.description}
          />
          <div className={styles.content}>
            {token ? (
              <InfoBlockWidget
                label="Amount"
                labelClassName={styles.infoLabel}
                value={
                  <TokenWidget
                    icon={token.icon}
                    noIcon
                    symbol={token.symbol}
                    amount={amount}
                    decimals={token.decimals}
                  />
                }
              />
            ) : (
              <div className={styles.loaderWrapper}>
                <LoadingIndicator />
              </div>
            )}

            <InputWrapper
              className={styles.detailsItem}
              labelClassName={styles.inputLabel}
              fieldName="gas"
              label="TGas"
            >
              <div className={styles.row}>
                <Input
                  className={cn(styles.inputWrapper, styles.detailsInput, {
                    [styles.error]: methods.formState.errors?.gas,
                  })}
                  inputStyles={{
                    width: getInputWidth(),
                  }}
                  onClick={e => e.stopPropagation()}
                  type="number"
                  min={0.1}
                  step={0.01}
                  max={0.3}
                  isBorderless
                  size="block"
                  placeholder={`${DEFAULT_PROPOSAL_GAS}`}
                  {...methods.register('gas')}
                />
              </div>
            </InputWrapper>

            {status === BountyStatus.InProgress ||
            status === BountyStatus.Expired ||
            status === BountyStatus.PendingApproval ? (
              <InfoBlockWidget
                labelClassName={styles.infoLabel}
                label="Claimed by"
                value={content.accountId}
              />
            ) : (
              <InfoBlockWidget
                label="Claims"
                labelClassName={styles.infoLabel}
                value={
                  <div>
                    <span>{content.slots}</span>
                    <span className={styles.slotsTotal}>
                      /{content.slotsTotal}
                    </span>
                  </div>
                }
              />
            )}
          </div>
        </div>
        {renderActionBar()}
      </form>
    </FormProvider>
  );
};
