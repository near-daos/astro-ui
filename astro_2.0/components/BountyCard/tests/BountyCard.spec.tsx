/* eslint-disable @typescript-eslint/ban-ts-comment, @typescript-eslint/no-empty-function */
import get from 'lodash/get';
import { render } from 'jest/testUtils';

import { Token } from 'types/token';
import { Proposal } from 'types/proposal';
import { BountyStatus } from 'types/bounties';

import { BountyCard } from 'astro_2.0/components/BountyCard';
import { BountyCardContent } from 'astro_2.0/components/BountyCard/types';

describe('bounty card', () => {
  const dao = {
    legal: {},
    id: 'myDaoId',
    txHash: '4px581fv2HJ5LuWBUs72qouAJnZ5ka4bEA2ZLmixgjHD',
    name: 'saturn',
    description:
      'Weâ€™re a community grant for artists who want to build projects on our platform. Join our Discord channel to stay up to date with latest info!',
    members: 3,
    daoMembersList: ['jason.born'],
    activeProposalsCount: 11,
    totalProposalsCount: 15,
    totalDaoFunds: 77,
    lastProposalId: 12,
    proposals: 13,
    totalProposals: 15,
    logo:
      'https://image.freepik.com/free-photo/blue-liquid-marble-background-abstract-flowing-texture-experimental-art_53876-104502.jpg',
    funds: '17043.60259',
    createdAt: '2021-10-22T12:46:32.885Z',
    groups: [
      {
        members: ['anima.testnet'],
        name: 'Council',
        permissions: [
          '*:Finalize',
          '*:AddProposal',
          '*:VoteApprove',
          '*:VoteReject',
          '*:VoteRemove',
        ],
        votePolicy: {},
        slug: 'Council',
      },
    ],
    policy: {
      createdAt: '2021-10-22T12:46:32.885Z',
      daoId: 'saturn.sputnikv2.testnet',
      proposalBond: '100000000000000000000000',
      bountyBond: '100000000000000000000000',
      proposalPeriod: '604800000000000',
      bountyForgivenessPeriod: '604800000000000',
      defaultVotePolicy: {
        weightKind: 'RoleWeight',
        quorum: '0',
        kind: 'Ratio',
        weight: '',
        ratio: [3, 10],
      },
      roles: [],
    },
    links: ['example.com'],
    displayName: 'Saturn',
  };

  const content = ({
    id: 3,
    daoId: 'kaleinik-token-test.sputnikv2.testnet',
    amount: '100',
    description: 'DIY',
    externalUrl: '',
    forgivenessPeriod: '604800000000000',
    bountyBond: '100000000000000000000000',
    status: 'Available',
    type: 'Bounty',
    timeToComplete: 'in 4 days',
    slots: 1,
    slotsTotal: 1,
    claimedByCurrentUser: false,
  } as unknown) as BountyCardContent;

  it('Should render component', () => {
    const { container } = render(
      <BountyCard
        content={content}
        showActionBar
        canClaim
        bountyId="1"
        dao={dao}
        deadlineThreshold="1"
        completeHandler={() => {}}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('Should render "due on" if bounty is not available', () => {
    const component = render(
      <BountyCard
        content={{
          ...content,
          status: BountyStatus.InProgress,
        }}
        showActionBar
        canClaim
        bountyId="1"
        dao={dao}
        deadlineThreshold="1"
        completeHandler={() => {}}
      />
    );

    expect(component.getAllByText('Due on', { exact: false })).toHaveLength(1);
  });

  it('Should render token info if available', () => {
    const component = render(
      <BountyCard
        content={{
          ...content,
          amount: '123456',
          token: ({
            id: '1',
            decimals: 1,
          } as unknown) as Token,
        }}
        showActionBar={false}
        canClaim
        bountyId="1"
        dao={dao}
        deadlineThreshold="1"
        completeHandler={() => {}}
      />
    );

    expect(component.getAllByText('12345.6')).toHaveLength(1);
  });

  it('Should render link to related proposal', () => {
    const component = render(
      <BountyCard
        content={{
          ...content,
          status: BountyStatus.PendingApproval,
        }}
        showActionBar
        relatedProposal={
          ({
            id: 'proposalId',
            daoId: 'myDaoId',
          } as unknown) as Proposal
        }
        canClaim
        bountyId="1"
        dao={dao}
        deadlineThreshold="1"
        completeHandler={() => {}}
      />
    );

    const linkComponent = component.getByText('Link to Proposal');

    expect(linkComponent).toBeDefined();
    expect(get(linkComponent, 'href')).toContain(
      'dao/myDaoId/proposals/proposalId'
    );
  });
});
/* eslint-enable @typescript-eslint/ban-ts-comment, @typescript-eslint/no-empty-function */
