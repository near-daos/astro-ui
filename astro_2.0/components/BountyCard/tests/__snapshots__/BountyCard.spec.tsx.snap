// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bounty card Should render component 1`] = `
<div>
  <form>
    <div>
      <div
        class=""
      >
        <div>
          <div
            class=""
          >
            Type
          </div>
        </div>
        <div
          class=""
        >
          <div
            class="undefined"
          >
            Bounty
          </div>
        </div>
      </div>
      <div>
        Complete in in 4 days
      </div>
      <div
        class=""
      >
        <div>
          <div
            class=""
          >
            Status
          </div>
        </div>
        <div
          class=""
        >
          <div
            class="undefined"
          >
            <div
              class=""
            >
              Available
            </div>
          </div>
        </div>
      </div>
      <div>
        <div>
          Description
        </div>
        <div>
          DIY
        </div>
        
      </div>
      <div>
        <div>
          <div
            class=""
          >
            <svg>
              <use
                href="/_next/undefined"
              />
            </svg>
          </div>
        </div>
        <div
          class=""
        >
          <div
            class=""
          >
            <span>
              TGas
            </span>
          </div>
          <div>
            <label
              class="undefined"
              for="1"
            >
              <input
                class=""
                id="1"
                max="0.3"
                min="0.1"
                name="gas"
                placeholder="0.1"
                step="0.01"
                style="text-align: left; width: 6ch;"
                type="number"
              />
            </label>
          </div>
        </div>
        <div
          class=""
        >
          <div>
            <div
              class=""
            >
              Claims
            </div>
          </div>
          <div
            class=""
          >
            <div
              class=""
            >
              <div>
                <span>
                  1
                </span>
                <span>
                  /
                  1
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class=""
      >
        <div>
          <div
            class=""
          >
            Bond
          </div>
          <div>
            <svg
              class=""
            >
              <use
                href="/_next/undefined"
              />
            </svg>
          </div>
        </div>
        <div
          class=""
        >
          <div
            class=""
          >
            <div>
              <div>
                0.1
              </div>
              <div>
                NEAR
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class=""
      >
        <div>
          <div
            class=""
          >
            Grace
          </div>
          <div>
            <svg
              class=""
            >
              <use
                href="/_next/undefined"
              />
            </svg>
          </div>
        </div>
        <div
          class=""
        >
          <div
            class=""
          >
            <div>
              <div>
                7
              </div>
              <div>
                days
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <button
          class=""
          type="submit"
        >
          Claim
        </button>
      </div>
    </div>
  </form>
</div>
`;
